crates/miden-lib/src/transaction/inputs.rs

```diff
@@ -178,251 +27,407 @@ enum AccountInputsType {
-///
-/// The data above is processed by `prologue::process_input_notes_data`.
-fn add_input_notes_to_advice_inputs(
-    tx_inputs: &TransactionInputs,
-    tx_args: &TransactionArgs,
-    inputs: &mut AdviceInputs,
-) -> Result<(), TransactionInputError> {
-    // if there are no input notes, nothing is added to the advice inputs
-    if tx_inputs.input_notes().is_empty() {
-        return Ok(());
+    /// Extends the stack with the given elements.
+    pub fn extend_stack(&mut self, iter: impl IntoIterator<Item = Felt>) {
+        self.0.stack.extend(iter);
     }
 
-    let mut note_data = Vec::new();
-    for input_note in tx_inputs.input_notes().iter() {
-        let note = input_note.note();
-        let assets = note.assets();
-        let recipient = note.recipient();
-        let note_arg = tx_args.get_note_args(note.id()).unwrap_or(&EMPTY_WORD);
-
-        // NOTE: keep map in sync with the `note::get_inputs` API procedure
-        inputs.extend_map([(
-            recipient.inputs().commitment(),
-            recipient.inputs().format_for_advice(),
-        )]);
+    /// Extends the [`MerkleStore`](miden_objects::crypto::merkle::MerkleStore) with the given
+    /// nodes.
+    pub fn extend_merkle_store(&mut self, iter: impl Iterator<Item = InnerNodeInfo>) {
+        self.0.store.extend(iter);
+    }
+
+    /// Extend the advice stack with the transaction inputs.
+    ///
+    /// The following data is pushed to the advice stack:
+    ///
+    /// [
+    ///     PARENT_BLOCK_COMMITMENT,
+    ///     PARTIAL_BLOCKCHAIN_COMMITMENT,
+    ///     ACCOUNT_ROOT,
+    ///     NULLIFIER_ROOT,
+    ///     TX_COMMITMENT,
+    ///     TX_KERNEL_COMMITMENT
+    ///     PROOF_COMMITMENT,
+    ///     [block_num, version, timestamp, 0],
+    ///     NOTE_ROOT,
+    ///     kernel_version
+    ///     [account_id, 0, 0, account_nonce],
+    ///     ACCOUNT_VAULT_ROOT,
+    ///     ACCOUNT_STORAGE_COMMITMENT,
+    ///     ACCOUNT_CODE_COMMITMENT,
+    ///     number_of_input_notes,
+    ///     TX_SCRIPT_ROOT,
+    /// ]
+    fn build_stack(
+        &mut self,
+        tx_inputs: &TransactionInputs,
+        tx_script: Option<&TransactionScript>,
+        kernel_version: u8,
+    ) {
+        let header = tx_inputs.block_header();
+
+        // ── block header data (keep in sync with kernel) ────────────────────
+        self.extend_stack(header.prev_block_commitment());
+        self.extend_stack(header.chain_commitment());
+        self.extend_stack(header.account_root());
+        self.extend_stack(header.nullifier_root());
+        self.extend_stack(header.tx_commitment());
+        self.extend_stack(header.tx_kernel_commitment());
+        self.extend_stack(header.proof_commitment());
+        self.extend_stack([
+            header.block_num().into(),
+            header.version().into(),
+            header.timestamp().into(),
+            ZERO,
+        ]);
+        self.extend_stack(header.note_root());
+
+        // kernel version
```

Comment:
I would format it as a new block the same way it is done for other data, i.e. 
```suggestion
        // ── kernel version ───────────────────────────────────────────────
```